addons/google_gmail/models/google_gmail_mixin.py

```diff
@@ -73,14 +150,30 @@ def _generate_oauth2_string(self, user, refresh_token):
         if not self.google_gmail_access_token \
            or not self.google_gmail_access_token_expiration \
            or self.google_gmail_access_token_expiration < now_timestamp:
-            self.google_gmail_access_token, expires_in = self.env['google.service']._get_access_token(
-                refresh_token, 'gmail', self._SERVICE_SCOPE)
-            self.google_gmail_access_token_expiration = now_timestamp + expires_in
 
-            _logger.info('Google Gmail: fetch new access token. Expire in %i minutes', expires_in // 60)
+            access_token, expiration = self._fetch_gmail_access_token(self.google_gmail_refresh_token)
+
+            self.write({
+                'google_gmail_access_token': access_token,
+                'google_gmail_access_token_expiration': expiration,
+            })
+
+            _logger.info('Google Gmail: fetch new access token. Expire in %i minutes',
+                (self.google_gmail_access_token_expiration - now_timestamp) // 60)
         else:
             _logger.info(
                 'Google Gmail: reuse existing access token. Expire in %i minutes',
                 (self.google_gmail_access_token_expiration - now_timestamp) // 60)
 
         return 'user=%s\1auth=Bearer %s\1\1' % (user, self.google_gmail_access_token)
+
+    def _get_gmail_csrf_token(self):
+        """Generate a CSRF token that will be verified in `google_gmail_callback`.
+
+        This will prevent a malicious person to make an admin user disconnect the mail servers.
+        """
+        self.ensure_one()
+        _logger.info('Google Gmail: generate CSRF token for %s #%i', self._name, self.id)
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        data = str((self.env.cr.dbname, 'google_gmail_oauth', self._name, self.id))
```

Comment:
message maybe ?

Reply:
message ? :thinking:

Reply:
The hmac tool in master talks about secret and message as variable name to make the hash (instead of "data").

Reply:
ah ok sorry,done